google/generativeai/types/content_types.py

```diff
@@ -240,15 +251,336 @@ def to_contents(contents: ContentsType) -> list[glm.Content]:
+                # ),
+                # 3. We support user-provided descriptions.
+                description=descriptions.get(name, None),
+            ),
+        )
+        for name, param in defaults.items()
+        # We do not support *args or **kwargs
+        if param.kind
+        in (
+            inspect.Parameter.POSITIONAL_OR_KEYWORD,
+            inspect.Parameter.KEYWORD_ONLY,
+            inspect.Parameter.POSITIONAL_ONLY,
+        )
+    }
+    parameters = pydantic.create_model(f.__name__, **fields_dict).schema()
+    # Postprocessing
+    # 4. Suppress unnecessary title generation:
+    #    * https://github.com/pydantic/pydantic/issues/1051
+    #    * http://cl/586221780
+    parameters.pop("title")
+    for name, function_arg in parameters.get("properties", {}).items():
+        function_arg.pop("title")
+        annotation = defaults[name].annotation
+        # 5. Nullable fields:
+        #     * https://github.com/pydantic/pydantic/issues/1270
+        #     * https://stackoverflow.com/a/58841311
+        #     * https://github.com/pydantic/pydantic/discussions/4872
+        if typing.get_origin(annotation) is typing.Union and type(None) in typing.get_args(
+            annotation
+        ):
+            function_arg["nullable"] = True
+    # 6. Annotate required fields.
+    if required:
+        # We use the user-provided "required" fields if specified.
+        parameters["required"] = required
+    else:
+        # Otherwise we infer it from the function signature.
+        parameters["required"] = [
+            k
+            for k in defaults
+            if (
+                defaults[k].default == inspect.Parameter.empty
+                and defaults[k].kind
+                in (
+                    inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                    inspect.Parameter.KEYWORD_ONLY,
+                    inspect.Parameter.POSITIONAL_ONLY,
+                )
+            )
+        ]
+    schema = dict(name=f.__name__, description=f.__doc__, parameters=parameters)
+    return schema
+
+
+def _rename_schema_fields(schema):
+    if schema is None:
+        return schema
 
+    schema = schema.copy()
 
-def to_tools(tools: ToolsType) -> list[glm.Tool]:
-    if tools is None:
-        return []
-    elif isinstance(tools, Mapping):
-        return [glm.Tool(tools)]
-    elif isinstance(tools, Iterable):
-        return [glm.Tool(t) for t in tools]
+    type_ = schema.pop("type", None)
+    if type_ is not None:
+        schema["type_"] = type_.upper()
+
+    format_ = schema.pop("format", None)
+    if format_ is not None:
+        schema["format_"] = format_
+
+    items = schema.pop("items", None)
+    if items is not None:
+        schema["items"] = _rename_schema_fields(items)
+
+    properties = schema.pop("properties", None)
+    if properties is not None:
+        schema["properties"] = {k: _rename_schema_fields(v) for k, v in properties.items()}
+
+    return schema
+
+
+class FunctionDeclaration:
```

Comment:
Since these classes are all exported, can you add docstrings? Just 1-liners at the class level would do, just enough for a user to understand at a glance.